File: Cyclic_Sort/CyclicSort.cpp
/*Implementation of Cyclic sort in C++. Works only with integers*/

#include <iostream>

using namespace std; 

void swapping(int &a, int &b) {     //swap the content of a and b
   int temp;
   temp = a;
   a = b;
   b = temp;
}

void display(int *array, int size) {  //display array
   for(int i = 0; i<size; i++)
      cout << array[i] << " ";
   cout << endl;
}

void cycleSort(int *array, int n) {
    
   for(int start = 0; start<n-1; start++) {    
      
      int key = array[start];
      int location = start;

      for(int i = start+1; i<n; i++) { //count smaller element in the right side of key
         if(array[i] < key)
            location++;
      }

      if(location == start) //when it is in correct place go for next iteration
         continue;
    
      while(key == array[location]) //if same data is found increase location
         location++;
         
      if(location != start)
         swapping(array[location], key);

      while(location != start) {
         location = start;

         for(int i = start+1; i<n; i++) { //location to put element
            if(array[i] < key)
               location++;
         }

         while(key == array[location]) //if same data is found increase location
            location++;
         if(key != array[location])
            swapping(key, array[location]);
      }
   }
}

//Driver programm
int main()
{
   int n;
   cout << "Enter the number of elements: ";
   cin >> n;
   int *arr = new int[n]; //create an array with given number of elements
   cout << "Enter elements : ";

   for(int i = 0; i<n; i++) {
      cin >> arr[i];
   }

   cout << "Array before Sorting: ";
   display(arr, n);
   cycleSort(arr, n);
   cout << "Array after Sorting: ";
   display(arr, n);

   return 0;
}

File: Cyclic_Sort/CyclicSort.py
def cyclicSort(array):
  writes = 0
   
  for cyclicStart in range(0, len(array) - 1):
    item = array[cyclicStart]
     
    pos = cyclicStart
    for i in range(cyclicStart + 1, len(array)):
      if array[i] < item:
        pos += 1
     
    if pos == cyclicStart:
      continue
     
    while item == array[pos]:
      pos += 1
    array[pos], item = item, array[pos]
    writes += 1
     
    while pos != cyclicStart:
       
      pos = cyclicStart
      for i in range(cyclicStart + 1, len(array)):
        if array[i] < item:
          pos += 1
       
      while item == array[pos]:
        pos += 1
      array[pos], item = item, array[pos]
      writes += 1
   
  return writes
   
arr = [1, 8, 3, 9, 10, 10, 2, 4 ]
n = len(arr)
cyclicSort(arr)
 
print("After sort : ")
for i in range(0, n) :
    print(arr[i], end = ' ')
